lib/include/jxl/encode.h

```diff
@@ -66,11 +66,49 @@ typedef enum {
    */
   JXL_ENC_NEED_MORE_OUTPUT = 2,
 
-  /** The encoder doesn't (yet) support this.
+} JxlEncoderStatus;
+
+/**
+ * Error conditions:
+ * API usage errors have the 0x80 bit set to 1
+ * Other errors have the 0x80 bit set to 0
+ */
+typedef enum {
+  /** No error
    */
-  JXL_ENC_NOT_SUPPORTED = 3,
+  JXL_ENC_ERR_OK = 0,
 
-} JxlEncoderStatus;
+  /** Generic encoder error due to unspecified cause
+   */
+  JXL_ENC_ERR_GENERIC = 1,
+
+  /** Out of memory
+   *  TODO(jon): actually catch this and return this error
+   */
+  JXL_ENC_ERR_OOM = 2,
+
+  /** JPEG bitstream reconstruction data could not be
+   *  represented (e.g. too much tail data)
+   */
+  JXL_ENC_ERR_JBRD = 3,
+
+  /** Input is invalid (e.g. corrupt JPEG file or ICC profile)
+   */
+  JXL_ENC_ERR_BAD_INPUT = 4,
+
+  /** The encoder doesn't (yet) support this. Either no version of libjxl
+   * supports this, and the API is used incorrectly, or the libjxl version
+   * should have been checked before trying to do this.
+   */
+  JXL_ENC_ERR_NOT_SUPPORTED = 0x80,
```

Comment:
I think there isn't anything you can programatically do to handle this error, so it's not needed to be able to handle it and it could be merged into JXL_ENC_ERR_GENERIC.

It would best be avoided to rely on encoder error values, except for a very specific programmatic use case, such as jbrd. This because if users of the API do rely on many different possible error values for behavior, this behavior needs to maintained in the library. For many errors, there's nothing you can programmatically do to recover from them, any API usage error shouldn't be returned, but can be printed in the console when compiling with the debug options only.

It could also be argued that JXL_ENC_ERR_BAD_INPUT is an API usage error, and checking of their validity should be done elsewhere.

WDYT?